app/android/src/main/kotlin/activity/MainActivity.kt

```diff
@@ -103,11 +148,158 @@ class MainActivity : AniComponentActivity() {
+            val torrentStorage = mediaCacheManager.storagesIncludingDisabled
+                .find { it is DataStoreMediaCacheStorage && it.engine is TorrentMediaCacheEngine }
+
+            if (torrentStorage == null) {
+                logger.error("[migration] Failed to get TorrentMediaCacheEngine, it is null.")
+                withContext(Dispatchers.Main) {
+                    migrationStatus =
+                        MigrationStatus.Error(IllegalStateException("Media cache storage with engine TorrentMediaCacheEngine is not found."))
+                }
+                return@launch
+            }
+            logger.info { "[migration] New torrent dir: $newPath" }
+
+            withContext(Dispatchers.Main) { migrationStatus = MigrationStatus.Metadata }
+            metadataStore.updateData { original ->
+                val nonTorrentMetadata = original.filter { it.engine != torrentStorage.engine.engineKey }
+                val torrentMetadata = original.filter { it.engine == torrentStorage.engine.engineKey }
+
+                nonTorrentMetadata + torrentMetadata.map { save ->
+                    save.copy(
+                        metadata = torrentStorage.engine
+                            .modifyMetadataForMigration(save.metadata, newPath.path),
+                    )
+                }
+            }
+
+            settingsRepo.mediaCacheSettings.update { copy(saveDir = newPath.absolutePath) }
+            logger.info { "[migration] Migration success." }
+
+            appTerminator.exitApp(this@MainActivity, 0)
+        } catch (e: Exception) {
+            withContext(Dispatchers.Main) { migrationStatus = MigrationStatus.Error(e) }
+            logger.error(e) { "[migration] Failed to migrate torrent cache." }
+        }
+    }
+
+    private sealed interface MigrationStatus {
+        object Init : MigrationStatus
+
+        class Cache(val currentFile: String?) : MigrationStatus
+
+        object Metadata : MigrationStatus
+
+        class Error(val throwable: Throwable? = null) : MigrationStatus
+    }
+
+    @Composable
+    private fun MigrationDialog(
+        status: MigrationStatus,
+    ) {
+        AlertDialog(
+            title = { Text("迁移缓存") },
+            text = {
+                Column {
+                    Text(renderMigrationStatus(status = status))
+                    if (status !is MigrationStatus.Error) {
+                        Spacer(modifier = Modifier.height(24.dp))
+                        LinearProgressIndicator(modifier = Modifier.fillMaxWidth())
+                        Spacer(modifier = Modifier.height(24.dp))
+                        Text(
+                            """
+                            从 4.9 版本起，Ani 不再将 BT 缓存存放之内部目录. 目前你的缓存位置在内部目录，Ani 正在迁移它们到外部目录.
+                            
+                            将缓存存放在外部目录可以使其他较高权限的应用访问或播放你的缓存，例如使用 Shizuku 授权的应用.
+                            
+                            此过程是完全自动的，迁移过程中设备可能会轻微卡顿. 请不要强制关闭 Ani，这可能导致缓存损坏或下次启动应用闪退.
```

Comment:
没必要说这么详细, 因为绝大部分用户都是不在意缓存到底放在哪的. 缓存的位置信息应该放在设置里, 让有需求的人自主看

```suggestion
                            此过程需要数秒钟，请不要强制关闭软件或手机。
```

Reply:
现在手机的存储速度是极快的，缓存文件顺序读写 1GB/s，一季度 3GB=3 秒，十个季度才 30 秒。不可能花十几分钟。

要保守一点可以：

```
此过程需要几分钟，请不要强制关闭软件或手机。
```

解释太多会导致用户看得不明白，期望比较少的人会懂 Shizuku。“较高权限”也不是一个通常人会懂的概念。